bin/jsii-release-maven

```diff
@@ -39,151 +41,182 @@ set -eu # we don't want "pipefail" to implement idempotency
-export GNUPGHOME=$(mktemp -d)
-export GPG_TTY=$(tty)
+    echo "Importing GPG key..." >&2
 
+    # GnuPG will occasionally bail out with "gpg: <whatever> failed: Inappropriate ioctl for device", the following attempts to fix
+    export GNUPGHOME=$(mktemp -d)
+    export GPG_TTY=$(tty)
 
-if [ -n "${MAVEN_GPG_PRIVATE_KEY}" ]; then
-    MAVEN_GPG_PRIVATE_KEY_FILE="${GNUPGHOME}/private.pem"
-    echo -e "${MAVEN_GPG_PRIVATE_KEY}" > ${MAVEN_GPG_PRIVATE_KEY_FILE}
-fi
 
-gpg --allow-secret-key-import --batch --yes --no-tty --import "${MAVEN_GPG_PRIVATE_KEY_FILE}" || {
-    echo "❌ GPG key import failed"
-    exit 1
+    if [ -n "${MAVEN_GPG_PRIVATE_KEY-}" ]; then
+        MAVEN_GPG_PRIVATE_KEY_FILE="${GNUPGHOME}/private.pem"
+        echo -e "${MAVEN_GPG_PRIVATE_KEY}" > ${MAVEN_GPG_PRIVATE_KEY_FILE}
+    fi
+
+    gpg --allow-secret-key-import --batch --yes --no-tty --import "${MAVEN_GPG_PRIVATE_KEY_FILE}" || {
+        error "GPG key import failed"
+    }
+
+    gpg_key_id=$(gpg --list-keys --with-colons | grep pub | cut -d: -f5)
+    echo "gpg_key_id=${gpg_key_id}"
+
+    GPG_PASSPHRASE_FROM_STDIN="--passphrase-fd 0"
+    if [[ "$(uname)" == "Darwin" ]]; then
+        # On Mac, we must pass this to disable a prompt for
+        # passphrase, but option is not recognized on Linux.
+        GPG_PASSPHRASE_FROM_STDIN="${GPG_PASSPHRASE_FROM_STDIN} --pinentry-mode loopback"
+    fi
+    export GPG_PASSPHRASE_FROM_STDIN
 }
 
-gpg_key_id=$(gpg --list-keys --with-colons | grep pub | cut -d: -f5)
-echo "gpg_key_id=${gpg_key_id}"
-
-GPG_PASSPHRASE_FROM_STDIN="--passphrase-fd 0"
-if [[ "$(uname)" == "Darwin" ]]; then
-    # On Mac, we must pass this to disable a prompt for
-    # passphrase, but option is not recognized on Linux.
-    GPG_PASSPHRASE_FROM_STDIN="${GPG_PASSPHRASE_FROM_STDIN} --pinentry-mode loopback"
-fi
-export GPG_PASSPHRASE_FROM_STDIN
-
-poms="$(find . -name '*.pom')"
-if [ -z "${poms}" ]; then
-    echo "❌ No JARS to publish: no .pom files found under $PWD"
-    exit 1
-fi
-
-echo "📦 Publishing to Maven Central"
-staging=$(mktemp -d)
-workdir=$(mktemp -d)
-
-# Create a settings.xml file with the user+password for maven
-mvn_settings="${workdir}/mvn-settings.xml"
-cat > ${mvn_settings} <<-EOF
+create_maven_settings() {
+    # Create a settings.xml file with the user+password for maven
+    mvn_settings="${workdir}/mvn-settings.xml"
+    if [[ "${is_signed}" == true ]]; then
+        cat > ${mvn_settings} <<-EOF
+<?xml version="1.0" encoding="UTF-8" ?>
+<settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
+                            http://maven.apache.org/xsd/settings-1.0.0.xsd">
+  <servers>
+    <server>
+      <id>${server_id}</id>
+      <username>\${env.MAVEN_USERNAME}</username>
```

Comment:
What's the rational of using `env.` here instead of just generating the file with the correct content? Is there any value in doing the substitution within maven?

Reply:
I got the idea from seeing how the setup-java GitHub action helps with deploys. I also think it's better to not be putting plain text credentials in a file.

Reply:
Okay, fair enough